homeassistant/components/color_extractor/__init__.py

```diff
@@ -0,0 +1,150 @@
+    SERVICE_TURN_ON as LIGHT_SERVICE_TURN_ON,
+)
+from homeassistant.helpers import aiohttp_client
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+# Extend the existing light.turn_on service schema
+SERVICE_SCHEMA = cv.make_entity_service_schema(
+    {
+        **LIGHT_TURN_ON_SCHEMA,
+        vol.Exclusive(ATTR_PATH, "color_extractor"): cv.isfile,
+        vol.Exclusive(ATTR_URL, "color_extractor"): cv.url,
+    }
+)
+
+
+def _get_file(file_path):
+    """Get a PIL acceptable input file reference.
+
+    Allows us to mock patch during testing to make BytesIO stream.
+    """
+    return file_path
+
+
+def _get_color(file_handler) -> tuple:
+    """Given an image file, extract the predominant color from it."""
+    color_thief = ColorThief(file_handler)
+
+    # get_color returns a SINGLE RGB value for the given image
+    color = color_thief.get_color(quality=1)
+
+    _LOGGER.debug("Extracted RGB color %s from image", color)
+
+    return color
+
+
+async def async_setup(hass, hass_config):
+    """Set up services for color_extractor integration."""
+
+    async def async_handle_service(service_call):
+        """Decide which color_extractor method to call based on service."""
+        service_data = dict(service_call.data)
+
+        if ATTR_URL not in service_data and ATTR_PATH not in service_data:
+            _LOGGER.error("Missing either required %s or %s key", ATTR_URL, ATTR_PATH)
+            return
+
+        try:
+            if ATTR_URL in service_data:
+                image_type = "URL"
+                image_reference = service_data.pop(ATTR_URL)
+                color = await async_extract_color_from_url(image_reference)
+
+            elif ATTR_PATH in service_data:
+                image_type = "file path"
+                image_reference = service_data.pop(ATTR_PATH)
+                color = await hass.async_add_executor_job(
+                    extract_color_from_path, image_reference
+                )
+
+        except UnidentifiedImageError as ex:
+            _LOGGER.error(
+                "Bad image from %s '%s' provided, are you sure it's an image? %s",
+                image_type,
+                image_reference,
+                ex,
+            )
+            return
+
+        if color:
+            service_data[ATTR_RGB_COLOR] = color
+
+            await hass.services.async_call(
+                LIGHT_DOMAIN, LIGHT_SERVICE_TURN_ON, service_data, blocking=True
+            )
+
+    hass.services.async_register(
+        DOMAIN,
+        SERVICE_TURN_ON,
+        async_handle_service,
+        schema=SERVICE_SCHEMA,
+    )
+
+    async def async_extract_color_from_url(url):
+        """Handle call for URL based image."""
+        if not hass.config.is_allowed_external_url(url):
+            _LOGGER.error(
+                "External URL '%s' is not allowed, please add to 'allowlist_external_urls'",
+                url,
+            )
+            return
```

Comment:
Inconsistent return value. We need to `return None` explicitly if we return a value anywhere in the function.